src/EventSender.h

```diff
@@ -11,12 +13,75 @@ extern "C" {
 }
 #undef typeid
 
+#include "memory/gpdbwrappers.h"
+
 class UDSConnector;
 struct QueryDesc;
 namespace yagpcc {
 class SetQueryReq;
 }
 
+#include <cstdint>
+
+struct QueryKey {
+  int tmid;
+  int ssid;
+  int ccnt;
+  int nesting_level;
+  uintptr_t query_desc_addr;
+
+  bool operator==(const QueryKey &other) const {
+    return std::tie(tmid, ssid, ccnt, nesting_level, query_desc_addr) ==
+           std::tie(other.tmid, other.ssid, other.ccnt, other.nesting_level,
+                    other.query_desc_addr);
+  }
+
+  static void register_qkey(QueryDesc *query_desc, size_t nesting_level) {
+    query_desc->yagp_hooks_query_state =
+        (YagpHooksQueryState *)ya_gpdb::palloc0(sizeof(YagpHooksQueryState));
+    int32 tmid;
+    gpmon_gettmid(&tmid);
+    query_desc->yagp_hooks_query_state->tmid = tmid;
+    query_desc->yagp_hooks_query_state->ssid = gp_session_id;
+    query_desc->yagp_hooks_query_state->ccnt = gp_command_count;
+    query_desc->yagp_hooks_query_state->nesting_level = nesting_level;
+    query_desc->yagp_hooks_query_state->query_desc_addr = (uintptr_t)query_desc;
+  }
+
+  static QueryKey qdesc_to_qkey(QueryDesc *query_desc) {
```

Comment:
I guess you can rename qdesc_to_qkey to `from_qdesc`, or better just make it an explicit constructor.